### PR TITLE
[ADP-3155] Encode Allegra ledger specs

### PR DESCRIPTION
**Pull request by paolino**
- [x] encode types from allegra specification pdf
- [x] add Script.fine to shelley era types
- [x] factor out common functionality for the era related specs

ADP-3155

### CODE REVIEW

File: lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import Address(Addr, Addr_rwd, Ix, Slot);
+import Crypto(VKey, Sig, ScriptHash, Script, KeyHash_G);
+import PParams(Coin, Epoch, PParamsUpdate);
+import Delegation(DCert);
+import Script(Script^native_v2);
+
+{-----------------------------------------------------------------------------
+    Transactions
+    Mary spec, Figure 5
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine:11]:
```suggestion
    Shelley spec, Figure 10
    and Mary spec, Figure 5
```

as this is a combination of both figures.

File: lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+AuxiliaryDataHash   = _ ; --^ auxiliary data hash
+
+-- Derived types
+
+TxIn                = TxId × Ix; --^ transaction input
+TxOut               = Addr × Coin; --^ transaction output
+UTxO                = TxIn ↦ TxOut; --^ unspent tx outputs
+Wdrl                = Addr_rwd ↦ Coin; --^ reward withdrawal
+Metadata            = ℕ ↦ Metadatum; --^ metadata
+
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine:26]:
```suggestion
Script              = Script^native_ph2;
```

That's how the Mary figure tells it. 🤷

File: lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+TxWitness           =
+    { txwitsVKey    : VKey ↦ Sig  --^ VKey witnesses
+    , txwitsScript  : ScriptHash ↦ Script  --^ script witnesses
+    };
+
+Tx                  =
+    { txbody        : TxBody  --^ transaction body
+    , txWitness     : TxWitness --^  transaction witnesses
+    , txAD          : AuxiliaryData? --^  metadata
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine:54]:
```suggestion
    , txAD          : AuxiliaryData? --^ Auxiliary data
```

File: lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    };
+
+AuxiliaryData       =
+    { scripts       : ℙ Script^native_v2  --^ set of native scripts
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine:58]:
```suggestion
    { scripts       : ℙ Script --^ optional scripts
```

Better matches the figure in Mary.

**Review by HeinrichApfelmus — APPROVED** [review]
Excellent, thank you! 😊 Small requests on the `Allegra/Tx.fine` module to better match the figure in the Mary spec.

File: lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine (hunk 2 above)
Comment (paolino) [reply lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine:28]: Script              = Script^native_v2; you mean ?

Comment (HeinrichApfelmus) [reply lib/fine-types/test/data/Cardano/Ledger/Allegra/Tx.fine:28]: Yes. 😱 Sorry.

**Pull request merged**